### ISSUE TITLE
Support for GCxGC-MS data formats

### ISSUE THREAD

**Issue by nilshoffmann** [msdk/msdk#28]
Hi,
I am currently starting to work on a LECO ChromaTOF csv peak table import method and would like
to soon work on netCDF-based support for 2D GCxGC-MS data from ChromaTOF exports.
I do have some questions whether I need to watch out for some assumptions in the higher-level
layers of the software:

1. the csv peak table export format is freely configurable, but I have already developed a decent
importer for Maltcms that copes with the various configuration options. I would like to port that
part of the code and I am positive that the mapping to msdk datastructures will be manageable.
However, the format consists of sparse MS1 EI spectra, so I am not sure, whether I should implement
support for RawDataFile or for FeatureTable? Would be great, if you could point me to the right
direction there.

2. The ChromaTOF netCDF export format for GCxGC-MS (at least with ChromaTOF from about
 two years ago), only supported 1D RT data in the netCDF file. To reconstruct the 2D
retention time information, we implemented two strategies in Maltcms: a) user-provided parameters
for second column modulation time (usually 1 - 6 s) and reciprocal scan frequency (usually between 50 and 200 Hz),
aka scan rate or period, b) automatic estimation of second column modulation time by autocorrelation of portions of
the TIC signal and estimation of the scan rate from the differences in scan acquisition time (in netCDF).
Would you consider it practical to require the user to provide the parameters manually? Usually they know
the two parameters but the second approach could work almost automatically except for cases, where the estimation
values vary too much (the approach proved to be quite stable, but requires more computational resources).
Should I implement this as a separate module? Unfortunately, the exported netCDF files are indistinguishable
from ther 1D cousins based on the contained attributes alone.

Best regards,
Nils

**Comment by dyrlund** [comment]
Hi Nils

Great that you are working on a ChromaTOF import method. I will be happy to help you out as this is also on my wish list for MSDK.

First off, I think that we should make one generic CSV feature table import which can import any CSV table into the MSDK data model. The columns of the CSV file should – if possible – be mapped to the column names of ColumnName.java or the common columns available in MSDKObjectBuilder.java. The implementation can be done in a similar way to the mzTab import currently available.

And here are my comments to your questions:

1. I suggest that all the spectra in the csv file is added to a new RawDataFile which is then associated to the FeatureTable where the rest of the data is stored.

2. I’ve never really been working with netCDF files. So I might not be the best to answer this. But are these results you are storing in netCDF or raw spectra data. If it is the results then since the 2D RT is available in the CSV file, could the user be requested to select both a netCDF and a CSV file? Then you can get the true 2D RT information from the CSV file.

FYI then we are running ChromaTOF 4.50.8.0 (August 24, 2012) here.

Best,

Thomas

**Comment by dyrlund** [comment]
@nilshoffmann Have you done any progress on this? If not, then I will make a generic CSV importer during the next days and we can then adjust it to also work with the ChromatTOF CSV format. What do you think?

**Comment by nilshoffmann** [comment]
@dyrlund Yes, please go ahead with that. I will add the domain specific parts and utility methods and can also add a number of example files for 1D and 2D data from ChromaTOF.

**Comment by dyrlund** [comment]
I added my initial work on a generic CSV importer in pull request #76. The GCxGX test currently fails due to some minor bug which I will look into after Christmas. 

Currently, I map the input column names to those in [ColumnName] (https://github.com/msdk/msdk/blob/master/msdk-datamodel/src/main/java/io/github/msdk/datamodel/featuretables/ColumnName.java). Anything which can’t be matched will be stored as String data.

**Comment by dyrlund** [comment]
@nilshoffmann I fixed the GCxGC bug so that ChromatTOF txt files can be imported. Please have a look at it and let me know what you think. It would also be nice if you can add some more test files for both 1D and 2D data.

Currently, MSDK does not support the below column name and the corresponding data is therefore added as String data. Should we add some or all of the column names or what do you think?
- CAS
- Sample concentration
- Quant Masses
- UniqueMass
- Quant S/N
- Similarity
- Weight
- Classifications
- Type
- Spectra

Also, we should probably convert the data in the spectra column into a MsSpectrum.

**Comment by nilshoffmann** [comment]
@dyrlund I would recommend to keep the column names within the module for now and provide sensible means to map the content to chromatogram info and other, similar objects. I have supported the following column names in Maltcms: http://sourceforge.net/p/maltcms/code/ci/release-1.4.0/tree/maltcms-io-csv/src/main/java/maltcms/io/csv/chromatof/ChromaTOFParser.java 
I already have a parser for the Spectra format in ChromaTOF files, there may be a need to support different parsing modes though, as ChromaTOF supports separate or fused 2D RT information in the csv file format.

**Comment by nilshoffmann** [comment]
Please see the following pull request for an initial implementation:
https://github.com/msdk/msdk/pull/91
